mitodl/unified-ecommerce
# Add system slug field, add selectable view set

_This is intended to be reviewed after #96 and #97 - you should still be able to test this without a Traefik setup, though._

### What are the relevant tickets?

https://github.com/mitodl/hq/issues/3166
https://github.com/mitodl/hq/issues/3158

### Description (What does it do?)

This does (in increasing order of complexity):
* Fixes a bug in `generate_test_data` (the soft deletion code changed and this hadn't gotten the memo yet)
* Updates the IntegratedSystem model to have an auto-generating slug
* Adds `AuthVariegatedViewSet`, which allows for viewsets to swap serializers depending on the current user's session status (and includes tests for this)

### How can this be tested?

Automated tests should pass.

Without a Traefik setup (and PRs #96 and #97), testing a logged-in but not staff user by hand will be a bit difficult as you won't be able to log in anywhere (there's no login UI). You can test with an anonymous user, or you can set up a session for an admin user and then flip the bits manually in the user record to make them not privileged. 

1. Set up an instance - you can do this quickly by doing `migrate`, then `createsuperuser`, then `generate_test_data` so  you have some systems and products.
2. Log into the Django Admin as the admin user.
3. Navigate to the IntegratedSystems endpoint: http://ue.odl.local:8073/api/v0/meta/integrated_system/ You should see a relatively full compliment of system data here. You should have the normal create/edit controls from DRF.
4. Navigate to the Products endpoint: http://ue.odl.local:8073/api/v0/meta/product/ You should see complete data for these.  You should have the normal create/edit controls from DRF.
5. Log out.
6. Navigate to the IntegratedSystems endpoint again. You should now see systems, but with an abbreviated set of data, and you should not be able to make modifications.
7. Navigate to the Product endpoint again. You should still see the complete dataset here. (This does use the AuthVariegatedViewSet but both read and read/write serializers are the same.) You should not be able to make modifications.
